File: javascript/kindergarten-garden/kindergarten-garden.js
const DEFAULT_STUDENTS = [
  'Alice',
  'Bob',
  'Charlie',
  'David',
  'Eve',
  'Fred',
  'Ginny',
  'Harriet',
  'Ileana',
  'Joseph',
  'Kincaid',
  'Larry',
];

const PLANT_CODES = {
  G: 'grass',
  V: 'violets',
  R: 'radishes',
  C: 'clover',
};

export class Garden {
  constructor(diagram, students = DEFAULT_STUDENTS) {
	this.plantsDiagram = diagram.split('\n');
	this.students = students.sort();
  }

  plants(student) {
	let plantsList = [];
	
	let pos = this.students.indexOf(student) * 2;
	let code0 = this.plantsDiagram[0].charAt(pos);
	let code1 = this.plantsDiagram[0].charAt(pos + 1);
	let code2 = this.plantsDiagram[1].charAt(pos);
	let code3 = this.plantsDiagram[1].charAt(pos + 1);

	plantsList.push(PLANT_CODES[code0]);
	plantsList.push(PLANT_CODES[code1]);
	plantsList.push(PLANT_CODES[code2]);
	plantsList.push(PLANT_CODES[code3]);

 	return plantsList;
	
  }
}

File: c/kindergarten-garden/kindergarten_garden.c
#include "kindergarten_garden.h"

#include <string.h>

plant_t char_to_plant(char plant_label);
size_t student_to_index(const char *student);


plants_t plants(const char *diagram, const char *student) {

    size_t student_index = student_to_index(student) * 2;
    size_t diagram_length = strlen(diagram) / 2 + 1;

    return (plants_t){
        {
            char_to_plant(diagram[student_index]),
            char_to_plant(diagram[student_index + 1]),
            char_to_plant(diagram[student_index + diagram_length]),
            char_to_plant(diagram[student_index + diagram_length + 1])
        }
    };

}

size_t student_to_index(const char *student) {
    
    char* students[] = { 
        "Alice", "Bob", "Charlie", "David", 
        "Eve", "Fred", "Ginny", "Harriet", 
        "Ileana", "Joseph", "Kincaid", "Larry"
    };

    for (size_t i = 0; i < 12; i++) {
        if (strcmp(students[i], student) == 0) {
            return i;
        }
    }

    return -1;
}

plant_t char_to_plant(char plant_label) { 

    switch(plant_label) { 
        case 'C': 
            return CLOVER; 
        case 'G': 
            return GRASS; 
        case 'R': 
            return RADISHES; 
        case 'V': 
            return VIOLETS; 
    }

    return -1;
}
